client/src/main/java/io/pravega/client/netty/impl/RawClient.java

```diff
@@ -131,12 +137,25 @@ private void closeConnection(Throwable exceptionToInflightRequests) {
     }
 
     public <T extends Request & WireCommand> CompletableFuture<Reply> sendRequest(long requestId, T request) {
+        return sendRequest(requestId, request, DEFAULT_RAWCLIENT_REQUEST_TIMEOUT);
+    }
+
+    public <T extends Request & WireCommand> CompletableFuture<Reply> sendRequest(long requestId, T request,
+                                                                                  Duration timeout) {
         return connection.thenCompose(c -> {
             log.debug("Sending request: {}", request);
             CompletableFuture<Reply> reply = new CompletableFuture<>();
             synchronized (lock) {
                 requests.put(requestId, reply);
             }
+            CompletableFuture<Void> timer = Futures.futureWithTimeout(timeout, connectionFactory.getInternalExecutor());
+            Futures.onTimeout(timer, ex -> {
+                synchronized (lock) {
```

Comment:
can we log the fact that we are completing the future exceptionally due to a timeout? (this will help to debug the exact the request which failed with a timeout)

Reply:
got it!